LSM: Table{Index,Filter,Data}Schema

sentientwaffle opened the pull request:
(This PR has lots of lines changed, but for the most part they are just moved.)

### Motivation

The scrubber and `GridRepairQueue` will need to decode table index blocks.
(Specifically, they need to enumerate all referenced filter/data blocks in an index block).

But the scrubber and the `GridRepairQueue` are not specialized per-tree.

## Overview

Rather than switching between decoders based on the tree id (which would be complicated -- each `Table`/`Tree` is a different specialized type), encode the schema's parameters directly into the table index block.

Then the scrubber/repair-queue constructs the relevant decoder schema at runtime.
This change also lends itself well to backwards-compatible table layout changes.

(Aside: Functions that use `Key` are still specialized. The scrubber/repair-queue don't need these anyway.)

sentientwaffle committed:
LSM: TableIndexSchema

Scrubber/GridRepairQueue need to decode table index blocks.
(Specifically, they need to enumerate all referenced filter/data blocks in an index block).

Rather than switching between decoders based on the tree id (which would be complicated), encode the schema's parameters directly into the table index block.
Then the scrubber/repair-queue can construct the relevant schema at runtime.

sentientwaffle committed:
LSM: TableFilterSchema

sentientwaffle committed:
LSM: TableDataSchema

sentientwaffle commented on src/lsm/level_data_iterator.zig:
```diff
@@ -139,21 +140,19 @@ pub fn LevelTableValueBlockIteratorType(comptime Table: type, comptime Storage:
-                LevelTableValueBlockIterator,
-                "read",
-                read,
-            );
+            const it = @fieldParentPtr(LevelTableValueBlockIterator, "read", read);
```
(Just reformatting).

sentientwaffle commented on src/vsr.zig:
```diff
@@ -795,9 +795,9 @@ pub const Header = extern struct {
-        if (self.context != 0) return "context != 0";
         if (self.view != 0) return "view != 0";
         if (self.op == 0) return "op == 0"; // address ≠ 0
+        if (self.commit != 0) return "commit != 0";
```
The data this field used to hold is now in `context` (along with some others).

sentientwaffle commented on src/lsm/table.zig:
```diff
@@ -239,52 +243,22 @@ pub fn TableType(
+            .key_layout_size = layout.block_key_layout_size,
+            .value_count_max = layout.block_value_count_max,
+            .value_size = value_size,
+        });
```
These are the schemas which encode _new_ blocks, so they are still comptime-known.

sentientwaffle commented on src/lsm/table.zig:
```diff
@@ -361,7 +335,7 @@ pub fn TableType(
 
                         data.key_count,
-                        data.block_value_count_max,
+                        data.value_count_max,
```
Renamed this to be more consistent.

jorangreef commented on src/lsm/compaction.zig:
```diff
@@ -410,10 +411,12 @@ pub fn CompactionType(
             // TODO(jamii) This copy can be avoided if we bypass the cache.
             stdx.copy_disjoint(.exact, u8, compaction.index_block_a, index_block);
+
+            const index_a = TableIndexSchema.from_index_block(compaction.index_block_a);
```
This is great. 🚀 

```suggestion
            const schema_index_block_a = SchemaIndexBlock.from(compaction.index_block_a);
```

* Clarify that `schema_index_block_a` is a schema and not an index block.
* Shorten the `from` method name since the type name is clear.
* Use MSB order to put Schema first.

batiati replied:
It's great to clarify that it's a schema, not an index block. However, I think `Schema` may not be the MSB, e.g `filter_block` vs `filter_bytes` vs `filter_schema` and `data_block` vs `data_schema`.

batiati reviewed (commented):
We just need to settle the naming choice 🤔.
Apart from that, LGMT.

sentientwaffle committed:
LSM: Move block schemas table.zig → schema.zig

Motivation:
- This puts `schema.` at the start of the type name.
- This was a natural change since the table block decoders don't actually anything with the rest of `table.zig`.
- This will put all block decoders (and any future block decoders, whether or not they are table blocks – see below) together.

As part of unifying the manifest log, the I'll add the manifest log block here as well, which is why the file isn't named `table_schema.zig`.

sentientwaffle committed:
LSM: Rename from_*_block to from()

batiati reviewed (approved):
:+1:
